File: src/main/java/de/tectoast/emolga/utils/config/ConfigSource.java
package de.tectoast.emolga.utils.config;

import de.tectoast.emolga.commands.Command;
import de.tectoast.jsolf.JSONObject;

import java.util.function.Supplier;

public enum ConfigSource {
    JSON((gid, configName, key, defaultValue) -> {
        JSONObject conf = Command.getEmolgaJSON().createOrGetJSON("configuration").createOrGetJSON(gid).createOrGetJSON(configName);
        return conf.has(key) ? conf.get(key) : defaultValue.get();
    });

    private final ConfigMapper configMapper;

    ConfigSource(ConfigMapper configMapper) {
        this.configMapper = configMapper;
    }

    public Object get(long gid, String configName, String key, Supplier<Object> defaultValue) {
        return configMapper.get(gid, configName, key, defaultValue);
    }

    @FunctionalInterface
    private interface ConfigMapper {
        Object get(long gid, String configName, String key, Supplier<Object> defaultValue);
    }
}

File: src/main/kotlin/de/tectoast/emolga/utils/config/ConfigSource.kt
package de.tectoast.emolga.utils.config

import de.tectoast.emolga.utils.config.ConfigSource.ConfigMapper
import de.tectoast.emolga.utils.json.Configuration
import de.tectoast.emolga.utils.json.db
import kotlinx.coroutines.runBlocking
import org.litote.kmongo.eq
import java.util.function.Supplier

enum class ConfigSource(private val configMapper: ConfigMapper) {
    JSON(ConfigMapper { gid: Long, configName: String, key: String, defaultValue: Supplier<Any> ->
        runBlocking {
            db.configuration.findOne(Configuration::guild eq gid)?.data?.get(
                configName
            )?.get(key) ?: defaultValue.get()
        }
    });

    operator fun get(gid: Long, configName: String, key: String, defaultValue: Supplier<Any>): Any {
        return configMapper[gid, configName, key, defaultValue]
    }

    private fun interface ConfigMapper {
        operator fun get(gid: Long, configName: String, key: String, defaultValue: Supplier<Any>): Any
    }
}
